tls/_common/prf.py

```diff
@@ -0,0 +1,69 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+from __future__ import absolute_import, division, print_function
+
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.hmac import HMAC
+
+
+def _p_hash(hash_algorithm, secret, seed, output_length):
+    """
+    A seed expansion function that uses a single hash function to expand a
+    secret and seed into the number of bytes specified by output_length.
+    """
+    result = b""
+    i = 1
+    while len(result) < output_length:
+        h = HMAC(secret, hash_algorithm, default_backend())
+        h.update(_a(secret, hash_algorithm, i, seed))
+        h.update(seed)
+        result += h.finalize()
```

Comment:
@ashfall I'm also happy with the use of `bytearray`.  `bytearray`s are indeed [backed by dynamic arrays, like lists](https://github.com/python/cpython/blob/9f31cf9457acb2d0c74581cd959f94dc3e03b7d1/Objects/bytearrayobject.c#L223-L271) so the runtime will be amortized O(1) -- just as good as using a list!

For benchmarking, I can recommend [perf](https://github.com/haypo/perf).  Its `timeit` implementation reports a more meaningful statistical moment (the median) and it tries to smooth out scheduling issues by running the microbenchmark across many processes.  It does a bunch of other cool stuff, too.  Check it out!